api/desecapi/replication.py

```diff
@@ -0,0 +1,125 @@
+import os
+import subprocess
+from datetime import datetime, timedelta
+
+import dns.query
+import dns.zone
+from celery import shared_task
+
+from desecapi import models
+
+
+class ReplicationException(BaseException):
+
+    def __init__(self, message, **kwargs):
+        super().__init__(message)
+        for k, v in kwargs.items():
+            self.__setattr__(k, v)
+
+
+class GitRepositoryException(ReplicationException):
+    pass
+
+
+class UnsupportedZoneNameException(ReplicationException):
+    pass
+
+
+class Repository:
+
+    def __init__(self, path):
+        self.path = path
+
+    def _git(self, *args):
+        cmd = ['/usr/bin/git'] + list(args)
+        print('>>> ' + str(cmd))
+
+        with subprocess.Popen(
+                cmd,
+                bufsize=0,
+                cwd=self.path,
+                stderr=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+        ) as p:
+            rcode = p.wait()
+            stderr = p.stderr.read()
+            stdout = p.stdout.read()
+            try:
+                stderr, stdout = stderr.decode(), stdout.decode()
+            except UnicodeDecodeError:
+                GitRepositoryException('git stdout or stderr was not valid unicode!', cmd=cmd, stderr=stderr, stdout=stdout)
+
+        if rcode != 0:
+            raise GitRepositoryException('git returned nonzero error code', cmd=cmd, rcode=rcode, stdout=stdout, stderr=stderr)
+
+        print('\n'.join(['<<< ' + s for s in stdout.split('\n')]))
+        return stdout
+
+    def commit_all(self, msg=None):
+        self._git('add', '.')
+        status = self._git('status')
+        if 'nothing to commit' not in status:
```

Comment:
May always be false if a zone name contains this string. Can the same be achieved by inspecting the return code?

Reply:
I changed it to consider the output of `git diff --numstat`, which is empty iff there are no changes